hmap/rtw_hmap.c

```diff
@@ -50,37 +65,52 @@ int rtw_hmap_get(rtw_hmap *map, void *key, void *out_value) {
 }
 
 void rtw_hmap_insert(rtw_hmap *map, void *key, void *data) {
-    unsigned long index = map->hash_func(key);
-    rtw_hmap_elem_ *last = map->elements[index];
-    if (last == NULL) {
-        rtw_hmap_elem_ *item = (rtw_hmap_elem_ *)malloc(sizeof(rtw_hmap_elem_));
-        item->key = key;
+    unsigned long index = map->hash_func(key) % RTW_HMAP_BUCKET_SIZE_;
 
-        item->data = malloc(map->value_len);
-        memcpy(item->data, data, map->value_len);
+    rtw_hmap_elem_ *last = map->elements[index];
+    if (last == NULL) {        
+        unsigned char temp[(map->key_len + map->value_len) / sizeof(char)];
+        memcpy(temp, key, map->key_len);
+        memcpy(temp + map->key_len, data, map->value_len);
+        
+        rtw_vec_push_back(&map->keys_and_values, temp);
 
+        rtw_hmap_elem_ *item = (rtw_hmap_elem_ *)malloc(sizeof(rtw_hmap_elem_));
         item->next = NULL;
+        item->key = map->keys_and_values.data + (map->keys_and_values.len-1) * (map->key_len + map->value_len);
+        
         map->elements[index] = item;
+      
         return;
     }
     for (;;) {
-        if (map->comp_func(key, last->key) % RTW_HMAP_BUCKET_SIZE_) {
-            memcpy(last->data, data, map->value_len);
+        if (map->comp_func(key, last->key)) {
+            memcpy(last->key+map->key_len, data, map->value_len);
             return;
         }
         if (last->next == NULL)
             break;
         last = last->next;
     }
-    last->next = (rtw_hmap_elem_ *)malloc(sizeof(rtw_hmap_elem_));
-    last->next->data = malloc(map->value_len);
 
-    memcpy(last->next->data, data, map->value_len);
-    last->next->key = key;
-    last->next->next = NULL;
+
+    unsigned char temp[(map->key_len + map->value_len) / sizeof(char)];
+    memcpy(temp, key, map->key_len);
+    memcpy(temp + map->key_len, data, map->value_len);
+    
+    rtw_vec_push_back(&map->keys_and_values, temp);
+    rtw_hmap_elem_ *item = (rtw_hmap_elem_ *)malloc(sizeof(rtw_hmap_elem_));
+    item->next = NULL;
+    
+    item->key = map->keys_and_values.data + (map->keys_and_values.len-1) * (map->key_len + map->value_len);
+    map->elements[index]->next = item;
 }
 
 void rtw_hmap_free(rtw_hmap *map) {
+    
+    rtw_vec_clear(&map->keys_and_values, 1);
+    rtw_vec_shrink(&map->keys_and_values);
```

Comment:
This is unnecessary. Vector needs to be free'd. Otherwise there will be memory leaks.